### PR TITLE
feat(kad): implement common traits on `RoutingUpdate`

### PR DESCRIPTION
**Pull request by arsenron**
## Description
A few weeks ago when I was debugging my wrong setup with kademlia, I could not conveniently debug `RoutingUpdate` after adding an address to the DHT. It would be nice to have a few derivable traits on a public enum.

## Notes & open questions
Maybe we could make `Kademlia::add_address` return Result, where could specify the reason why adding address to the DHT failed. Something like `Result<Success/Pending, BucketFull/SelfEntry>`

## Change checklist

- [x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] A changelog entry has been made in the appropriate crates

### CODE REVIEW

File: protocols/kad/src/behaviour.rs [hunk 1]
```diff
@@ -3289,6 +3289,7 @@ impl fmt::Display for NoKnownPeers {
 impl std::error::Error for NoKnownPeers {}
 
 /// The possible outcomes of [`Kademlia::add_address`].
+#[derive(Debug, Clone)]
```
Comment (thomaseizinger) [inline protocols/kad/src/behaviour.rs]: Might as well add Copy and PartialEq too then?

**Review by thomaseizinger — APPROVED** [review]
LGTM

Can you add a changelog entry and bump the version? See https://github.com/libp2p/rust-libp2p/blob/master/CONTRIBUTING.md#write-changelog-entries-for-user-facing-changes.

**Comment by thomaseizinger** [comment]
> Maybe we could make `Kademlia::add_address` return Result, where could specify the reason why adding address to the DHT failed. Something like `Result<Success/Pending, BucketFull/SelfEntry>`

I am open to that but it will be a breaking change so we can't merge that immediately. Are you willing to submit a PR and explore the idea?

**Comment by arsenron** [comment]
> > Maybe we could make `Kademlia::add_address` return Result, where could specify the reason why adding address to the DHT failed. Something like `Result<Success/Pending, BucketFull/SelfEntry>`
> 
> I am open to that but it will be a breaking change so we can't merge that immediately. Are you willing to submit a PR and explore the idea?

Yeah, will look into it.

p.s. Updated changelog and added more derivable traits

File: protocols/kad/CHANGELOG.md [hunk 2]
```diff
@@ -1,4 +1,10 @@
-## 0.44.3 - unreleased
+## 0.44.4 - unreleased
+- Add a bunch of derivable traits to `RoutingUpdate`
+  See [PR 4270].
```
Comment (thomaseizinger) [inline protocols/kad/CHANGELOG.md]:
```suggestion

- Implement common traits on `RoutingUpdate`.
  See [PR 4270].
```

File: protocols/kad/CHANGELOG.md [hunk 3]
```diff
@@ -1,4 +1,11 @@
-## 0.44.3 - unreleased
+## 0.44.4 - unreleased
```
Comment (thomaseizinger) [inline protocols/kad/CHANGELOG.md:1]: You will also need to bump the version in `Cargo.toml` please! :)

Comment (arsenron) [reply protocols/kad/CHANGELOG.md:1]: Sure! Done.

**Comment by thomaseizinger** [comment]
I can't push to your branch unfortunately but you'll also have to update the lockfile and bump the version of `libp2p-kad` in the root manifest. See CI failures! :)

Alternatively, you can apply this patch:

```patch
Subject: [PATCH] Bump version in root manifest
---
Index: Cargo.lock
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/Cargo.lock b/Cargo.lock
--- a/Cargo.lock	(revision ae2b23056ea58467f8d322f789a62bdc8e06684f)
+++ b/Cargo.lock	(revision 87a70ef5a5db7c4bd1278959eae53cf7fdd9aa71)
@@ -2813,7 +2813,7 @@
 
 [[package]]
 name = "libp2p-kad"
-version = "0.44.3"
+version = "0.44.4"
 dependencies = [
  "arrayvec",
  "async-std",
Index: Cargo.toml
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/Cargo.toml b/Cargo.toml
--- a/Cargo.toml	(revision ae2b23056ea58467f8d322f789a62bdc8e06684f)
+++ b/Cargo.toml	(revision 87a70ef5a5db7c4bd1278959eae53cf7fdd9aa71)
@@ -73,7 +73,7 @@
 libp2p-gossipsub = { version = "0.45.0", path = "protocols/gossipsub" }
 libp2p-identify = { version = "0.43.0", path = "protocols/identify" }
 libp2p-identity = { version = "0.2.2" }
-libp2p-kad = { version = "0.44.3", path = "protocols/kad" }
+libp2p-kad = { version = "0.44.4", path = "protocols/kad" }
 libp2p-mdns = { version = "0.44.0", path = "protocols/mdns" }
 libp2p-metrics = { version = "0.13.1", path = "misc/metrics" }
 libp2p-mplex = { version = "0.40.0", path = "muxers/mplex" }
```

**Review by thomaseizinger — APPROVED** [review]

**Comment by mxinden** [comment]
@arsenron let us know in case you need the patch release cut rather sooner than later.

**Pull request merged**